File: SHM_13_Classes_with_XPointer.R
# In this scenario, compared to the 12th, we consider the case where the class
# does not own the data itself, but just an own pointer to them.
# ----------------------------
Rcpp::sourceCpp("SHM_13_Classes_with_XPointer.cpp")

# create external pointer obj to our matrix and project data to the heap
xptr_mat <- create_XPtr_for_R_obj(matrix(rnorm(12), 3, 4))

# check out the obj
xptr::xptr_address(xptr_mat)
print(xptr_mat)


# use external pointer as input to class obj creation where the
# the class owns a copy of the external pointer to the obj
external_mat_admin <- new(External_Matrix_Administration, xptr_mat)
external_mat_admin$print_Matrix()
external_mat_admin$add_42_to_Matrix()
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$add_number_to_Matrix(42.42)
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$add_number_to_Matrix(42L)
external_mat_admin$print_Matrix()
external_mat_admin$return_Matrix()
external_mat_admin$return_pointer_to_Data() # note the pointer address differs
                                            # due to copy of pointer
                                            
# object no longer needed: delete it, release pointers and memory
rm(external_mat_admin)

# check out the obj
xptr::xptr_address(xptr_mat)
print(xptr_mat)

# release memory
xptr::xptr_clear(xptr_mat)
xptr::xptr_address(xptr_mat)
print(xptr_mat)
# the memory is now released, delete pointer obj
rm(xptr_mat)

# NOTE: the External_Matrix_Administration objs is still running(!) if we delete
#       the external pointer obj and call the class obj again. The class owns a
#       copy of the external pointer.

File: SHM_13_Classes_with_XPointer.cpp
// [[Rcpp::depends(RcppArmadillo)]]
#include <iostream>
#include <RcppArmadillo.h>


class External_Matrix_Administration {
  
public:
  
  // ctor
  // --------------------
  External_Matrix_Administration (Rcpp::XPtr<arma::mat> xptr_A) {
    Rcpp::Rcout << xptr_A << std::endl; // the address coincides with R's xptr
    this->armaMat = arma::mat((double*) xptr_A->memptr(), xptr_A->n_rows, xptr_A->n_cols, false, false);
  };
  
  // dtor
  // --------------------
  ~External_Matrix_Administration() = default;
  
  // Member method
  // --------------------
  void print_Matrix() const{
    armaMat.print();
    Rcpp::Rcout << std::endl;
  }
  
  void add_42_to_Matrix() {
    armaMat += 42;
  }
  
  void add_number_to_Matrix(double number) {
    armaMat += number;
  }
  
  arma::mat return_Matrix() const{
    return armaMat;
  }
  
private:
  // Attribute
  // --------------------
  arma::mat armaMat;
};


Rcpp::XPtr<arma::mat> create_XPtr_for_R_obj(arma::mat& A) {
  arma::mat* armaMat = new arma::mat(A.memptr(), A.n_rows, A.n_cols, false, false);
  return Rcpp::XPtr<arma::mat> (armaMat);
}


RCPP_MODULE(mod_External_Matrix_Admin) {
  using namespace Rcpp;
  
  function("create_XPtr_for_R_obj",
           &create_XPtr_for_R_obj,
           List::create(_["mat"]), "Create external pointer of R matrix obj");
  
  
  Rcpp::class_<External_Matrix_Administration>("External_Matrix_Administration")
    .constructor<Rcpp::XPtr<arma::mat>>("Creates new class obj.")
    .method("print_Matrix", &External_Matrix_Administration::print_Matrix, "Prints matrix.")
    .method("add_42_to_Matrix", &External_Matrix_Administration::add_42_to_Matrix, "Add 42 to matrix.")
    .method("return_Matrix", &External_Matrix_Administration::return_Matrix, "Return matrix to R.")
    .method("add_number_to_Matrix", &External_Matrix_Administration::add_number_to_Matrix, "Add number to Matrix.")
  ;
  
}